### ISSUE TITLE
Make `resultEventRef` optional for Action EventRef definitions

### ISSUE THREAD

**Issue by ricardozanini** [serverlessworkflow/specification#567]
**What would you like to be added**:

This was already discussed in #161, but I'd like to bring it up again since it might be scenarios where I want to fire a function and not wait for a result. The spec can cover such use cases by making this field optional instead.

Additionally, this field should be renamed not to be confused with the `eventRef` attribute in other places, such as in the `callback` state. The new name can be `eventActionRef` or something like that.

**Why is this needed**:

1. [Make `resultEventRef` optional](https://github.com/serverlessworkflow/specification/blob/main/specification.md#EventRef-Definition) to implement fire and forget use cases or to call void functions
2. [Change the name of the `action.eventRef`](https://github.com/serverlessworkflow/specification/blob/main/specification.md#Action-Definition) to something else to not be confused with `eventRef` that's a string referencing an actual event definition.

I can send a PR once we settle it.

**Comment by fjtirado** [comment]
On top of that, it can be also argued why do we need resultEventRef at all, if fire&wait result can be achieved by using a callback state which action has a triggerEventRef.

**Comment by tsurdilo** [comment]
@ricardozanini 

1. imo resultEventRef is needed if "invoke" is set to "sync" (which is the default). That's the only way to know if the function invocation completed. If "invoke" is set to "async" then i think we can just say that resultEventRef can be ignored (not required). 
2. not sure this is needed to rename, as the other props are "functionRef" and "subflowRef", which all fits together atm. imo would be more confusing to rename eventRef but keep the other two unchanged.

**Comment by tsurdilo** [comment]
> @fjtirado On top of that, it can be also argued why do we need resultEventRef at all, if fire&wait result can be achieved by using a callback state which action has a triggerEventRef.

its needed when you want to invoke a function (sync, waiting on it to complete) inside your actions array in like operations state, or foreach state, or parallel state.  If you want async then just set "invoke" prop to "async".

**Comment by ricardozanini** [comment]
> imo resultEventRef is needed if "invoke" is set to "sync" (which is the default). That's the only way to know if the function invocation completed. If "invoke" is set to "async" then i think we can just say that resultEventRef can be ignored (not required). If you chose to wait on result ("invoke" is sync) and you don't want to merge the result back into state data, you can use action data filter to set that currently.

Having the `sync` option emphasizes the argument to make `resultEventRef` optional. What we can do is to make the `resultEventRef` required if `invoke` is `sync`. But this sounds marginally the same thing. We can get rid of the `invoke` attribute and just leave `resultEventRef` not required. If it's there the invocation is assumed to be sync.

 > Within callback state i think it also has value, you can send an event that triggers a service, in which case resultEventRef can be just an "ack" event and then still have the callback event that denotes the actual completion of the service invocations or chain of services, that may come days or weeks later.

I agree, but you can't assume that every use case will have an ack. To cover most of them, won't hurt making this attribute optional.

> not sure this is needed to rename, as the other props are "functionRef" and "subflowRef", which all fits together atm. imo would be more confusing to rename eventRef but keep the other two unchanged.

You're right, we can rethink these names as well as they appear in other attributes of the workflow definition. wdyt?

**Comment by tsurdilo** [comment]
> We can get rid of the invoke attribute and just leave resultEventRef not required. If it's there the invocation is assumed to be sync.

rather not have to rely on a property called "resultEventRef" to specify if this function is called sync or async, it's just not imo clear.
"invoke": "sync", "invoke": "async" is imo more clear as it is now

> I agree, but you can't assume that every use case will have an ack. To cover most of them, won't hurt making this attribute optional.

sure, we said to make it optional, so your runtime can ignore it if it wants

**Comment by tsurdilo** [comment]
Let's take a look at some examples like before/after and see what looks better, think that would give more perspective for this change to see what makes more sense or not. Can you please provide? 

Also we added async stuff in 0.8 and having it removed or completely changed in 0.9 or changed is maybe a little bit too risky. If there is a situation that current dsl cannot currently define, lets add that for sure, otherwise let's try to keep it to minimum changes and then try to make bigger changes if necessary in 0.10

**Comment by ricardozanini** [comment]
> rather not have to rely on a property called "resultEventRef" to specify if this function is called sync or async, it's just not imo clear.
"invoke": "sync", "invoke": "async" is imo more clear as it is now

I don't have a strong opinion about this. Either option is fine for me.

> Let's take a look at some examples like before/after and see what looks better, think that would give more perspective for this change to see what makes more sense or not. Can you please provide?

Sure I can create an example, but it's pretty straightforward. You create a callback with an event action that will only trigger the event and won't expect a response. I think what you're asking is a use case example. I'll provide one once I get back.

> sure, we said to make it optional, so your runtime can ignore it if it wants

That would be nice. 👍 

> Also we added async stuff in 0.8 and having it removed or completely changed in 0.9 is maybe a little bit too risky.

Oh I don't think this change will impact the async stuff added in 0.8. I might be missing something, thou.

**Comment by cdavernas** [comment]
I just saw that issue now, and this is related to discussions we had in private, @tsurdilo. I think what's needed here is what I asked for, meaning an action to produce a single event, and one to consume a single event. As stated before, this has imho nothing to do with sync/async patterns, but rather with eventing. 
Other solutions would imo be confusing, at best.

Wdyt?

**Comment by tsurdilo** [comment]
@cdavernas @ricardozanini 
Would be nice to have some example json/yaml to showcase what's being asked. Think that would clear things up for me as to what's actually requested. Maybe something like show whats currently available in 0.8 and what issue cannot be solved with it, and then the proposal on how to solve it. Thanks!!

**Comment by ricardozanini** [comment]
Hi Tiho!

Let's keep this simple for this issue. We can further discuss other topics like the fire and forget and renaming. The `resultEventRef` being optional will cover a simple use case where I have a Callback state that needs to fire an event and **not expect a response**. Such as:

```json
      {
        "name": "CheckPeerApprovalTask",
        "type": "callback",
        "action": {
          "eventRef": {
            "triggerEventRef": "notifyMyPeers",
            "data": ".transaction.data"
          }
        },
        "eventRef" : "TaskApproved",
        "end": true
      }
```

Does that make sense?

**Comment by tsurdilo** [comment]
@ricardozanini 
so remove resultEventRef from required props here: https://github.com/serverlessworkflow/specification/blob/main/schema/workflow.json#L583 ?
That fine. Let's go for it :)
